[JBPM-8293] Add Apache Spark prediction service example

ruivieira opened the pull request:
Unit tests are disabled, since they require creating a local Apache Spark cluster (driver, workers and Kie server in the same JVM).

gmunozfe commented on jbpm-recommendation-spark-random-forest/src/main/java/org/jbpm/prediction/randomforest/SparkRandomForest.java:
```diff
@@ -0,0 +1,213 @@
+                new StructField[] {
+                        DataTypes.createStructField("user", DataTypes.StringType, false),
+                        DataTypes.createStructField("level", DataTypes.IntegerType, false),
+                        DataTypes.createStructField("label", DataTypes.IntegerType, false),
```
Why this structure doesn't contain "item"? If "item" is not used for training the model, perhaps it may be removed from the readme and tests for clarity.

gmunozfe commented on jbpm-recommendation-spark-random-forest/src/main/java/org/jbpm/prediction/randomforest/SparkRandomForest.java:
```diff
@@ -0,0 +1,213 @@
+        predictionSchema = DataTypes.createStructType(
+                new StructField[] {
+                        DataTypes.createStructField("user", DataTypes.StringType, false),
+                        DataTypes.createStructField("level", DataTypes.IntegerType, false)
```
Same here

gmunozfe commented on jbpm-recommendation-spark-random-forest/src/main/java/org/jbpm/prediction/randomforest/SparkRandomForest.java:
```diff
@@ -0,0 +1,213 @@
+    public PredictionOutcome predict(Task task, Map<String, Object> inputData) {
+        Map<String, Object> outcomes = new HashMap<>();
+
+        if (pm != null) {
```
You may change the condition to positive and return immediately the new PredictionOutcome, without 'else' clause and one level less of indentation.
```suggestion
        if (pm == null)
            return new PredictionOutcome();
```

ruivieira commented:
@gmunozfe Thank you for the comments! A list of changes follows:

- Replaced `OneHotEncoder` (deprecated) with `OneHotEncoderEstimator`
- Number of trees as static final
- Rewrite `Boolean` to outcome to avoid NPE
- Removed dev comment
- Removed unused method (`assertBetween`)
- Added `item` variable to model
- Return `PredictionOutcome` immediately if pipeline model not initialised

The adding of `item` to the model implied the following changes:

- Before fitting the model, a number of unique `item` and `actor` values are needed. `MINIMUM_OBSERVATIONS` wasn't suited for this. Since Spark will throw an `IllegalArgumentException` in this case, we will check for it so that training can start when we have sufficient training data.
- Since we have more features, the position of the outcomes and probabilities moved up in the return prediction row.

Since the overhead for the training is high, it is now grouped in mini-batches of 20 data points. This doesn't change the results in the long run, but the running time for at least one of the test goes down considerably (from 30 minutes to 2 minutes in my case).

mswiderski commented:
@gmunozfe do you have any other comments or is this ready to be merged?

ruivieira commented:
@gmunozfe: as we discussed, I've removed the explicit Spark context variable (just using Spark session now) and also removed the `@PreDestroy` method. Thanks!
